From 7f8e00edb11b56be9810ca4946ba255c359c5a57 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C3=81lvaro=20Fern=C3=A1ndez=20Rojas?= <noltari@gmail.com>
Date: Sun, 14 Mar 2021 20:03:44 +0100
Subject: [PATCH] net: dsa: b53: add support for BCM63xx RGMIIs
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>
---
 drivers/net/dsa/b53/b53_common.c | 34 ++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

--- a/drivers/net/dsa/b53/b53_common.c
+++ b/drivers/net/dsa/b53/b53_common.c
@@ -1166,6 +1166,29 @@ static void b53_adjust_link(struct dsa_s
 			      tx_pause, rx_pause);
 	b53_force_link(dev, port, phydev->link);
 
+	if (is63xx(dev) && phy_interface_is_rgmii(phydev)) {
+		if (port == B53_CPU_PORT)
+			off = B53_RGMII_CTRL_IMP;
+		else
+			off = B53_RGMII_CTRL_P(port);
+
+		b53_read8(dev, B53_CTRL_PAGE, off, &rgmii_ctrl);
+		rgmii_ctrl &= ~(RGMII_CTRL_DLL_RXC | RGMII_CTRL_DLL_TXC |
+				RGMII_CTRL_ENABLE_GMII);
+		if (phydev->interface == PHY_INTERFACE_MODE_RGMII_ID)
+			rgmii_ctrl |= (RGMII_CTRL_DLL_RXC | RGMII_CTRL_DLL_TXC);
+		else if (phydev->interface == PHY_INTERFACE_MODE_RGMII_RXID)
+			rgmii_ctrl |= RGMII_CTRL_DLL_RXC;
+		else if (phydev->interface == PHY_INTERFACE_MODE_RGMII_TXID)
+			rgmii_ctrl |= RGMII_CTRL_DLL_TXC;
+		if (port != B53_CPU_PORT)
+			rgmii_ctrl |= RGMII_CTRL_ENABLE_GMII;
+		b53_write8(dev, B53_CTRL_PAGE, off, rgmii_ctrl);
+
+		dev_info(ds->dev, "Configured port %d for %s\n", port,
+			 phy_modes(phydev->interface));
+	}
+
 	if (is531x5(dev) && phy_interface_is_rgmii(phydev)) {
 		if (port == 8)
 			off = B53_RGMII_CTRL_IMP;
@@ -1357,6 +1380,25 @@ void b53_phylink_mac_link_up(struct dsa_
 			     bool tx_pause, bool rx_pause)
 {
 	struct b53_device *dev = ds->priv;
+	u8 rgmii_ctrl = 0, off;
+
+	if (is63xx(dev) && port >= 4) {
+		off = B53_RGMII_CTRL_P(port);
+
+		b53_read8(dev, B53_CTRL_PAGE, off, &rgmii_ctrl);
+		rgmii_ctrl &= ~(RGMII_CTRL_DLL_RXC | RGMII_CTRL_DLL_TXC);
+		if (phydev->interface == PHY_INTERFACE_MODE_RGMII_ID)
+			rgmii_ctrl |= (RGMII_CTRL_DLL_RXC | RGMII_CTRL_DLL_TXC);
+		else if (phydev->interface == PHY_INTERFACE_MODE_RGMII_RXID)
+			rgmii_ctrl |= RGMII_CTRL_DLL_RXC;
+		else if (phydev->interface == PHY_INTERFACE_MODE_RGMII_TXID)
+			rgmii_ctrl |= RGMII_CTRL_DLL_TXC;
+		rgmii_ctrl |= RGMII_CTRL_ENABLE_GMII;
+		b53_write8(dev, B53_CTRL_PAGE, off, rgmii_ctrl);
+
+		dev_info(ds->dev, "Configured port %d for %s\n", port,
+			 phy_modes(phydev->interface));
+	}
 
 	if (mode == MLO_AN_PHY)
 		return;
